feat: add sourceType: commonjs option

ota-meshi opened the pull request:
close #1376 

I think it's still under discussion how to design the option, but I tried implementing it.
I think the test cases included in this PR will be useful for discussion.

ota-meshi commented on test/tests-commonjs.js:
```diff
@@ -0,0 +1,91 @@
+
+// Illegal return statement with commonjs
+testFail(`class X { static { return; } }`, "'return' outside of function (1:19)", {ecmaVersion: 13, sourceType: "commonjs"});
+// with allowReturnOutsideFunction option
+test(`class X { static { return; } }`, {}, {allowReturnOutsideFunction: true, ecmaVersion: 13, sourceType: "commonjs"});
```
Remove this test due to changes in #1378.

```suggestion
```

RReverser commented:
This looks pretty good to me. The only other thing that comes to mind is it might be worth running all the `sourceType: "script"` tests in CommonJS mode as well, as for most of them it should be true.

ota-meshi commented:
Thank you for checking this PR!
I modified `test/run.js` to run all the `sourceType: "script"` tests and the same `commonjs` tests.
https://github.com/acornjs/acorn/pull/1377/commits/e610dd2c3b4ced4f64f887eca80991d69c2df38f
(Please let me know if this method should be avoided.)

After adding tests, I noticed that when using `commonjs` with `allowAwaitOutsideFunction`, the top-level `await` couldn't be parsed, so I fixed that.
But maybe `allowAwaitOutsideFunction` should be ignored for `commonjs`. What do you think about this?

ota-meshi commented:
Thank you for the comment! I changed it to reject the combination of `allowAwaitOutsideFunction` and `moduleType: "commonjs"`.
